src/app/reporting/Engine.cpp

```diff
@@ -99,17 +97,39 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
     // TODO: Need to handle multiple chunk of message
     while (clusterInfo != nullptr)
     {
-        if (clusterInfo->IsDirty())
+        if (apReadHandler->IsInitialReport())
+        {
+            // Retrieve data for this cluster instance and clear its dirty flag.
+            err = RetrieveClusterData(attributeDataList, *clusterInfo);
+            VerifyOrExit(err == CHIP_NO_ERROR,
+                         ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
+            attributeClean = false;
+        }
+        else
         {
-            if (apReadHandler->IsInitialReport())
+            ClusterInfo * path = mpChangePathList;
+            while (path != nullptr)
             {
-                // Retrieve data for this cluster instance and clear its dirty flag.
-                err = RetrieveClusterData(attributeDataList, *clusterInfo);
+                if (clusterInfo->IsAttributePathSupersetOf(*path))
+                {
+                    err = RetrieveClusterData(attributeDataList, *path);
+                }
+                else if (path->IsAttributePathSupersetOf(*clusterInfo))
+                {
+                    err = RetrieveClusterData(attributeDataList, *clusterInfo);
+                }
+                else
+                {
+                    // partial overlap is not possible, hence the 'continue' here: clusterInfo and path have nothing in
+                    // common.
+                    path = path->mpNext;
```

Comment:
You can remove this line, and the code will be more readable.

Reply:
updated